Fix: Align tooltip with others element spacing wise

Rishit30G opened the pull request:
## What?
This PR is intended to fix: https://github.com/WordPress/gutenberg/issues/64440

## How?
Added specific styles to `Move Up` and `Move Down` Labels

## Testing Instructions
- Create a new post 
- Add a paragraph 
- Notice the direction button label spacing from the Toobar

## Screenshots or screencast <!-- if applicable -->

https://github.com/user-attachments/assets/c3bf3597-5e0e-44d3-a076-5e9ded4a2d28

Rishit30G committed:
alter styles of move up and move down label

t-hamano reviewed (commented):
Thanks for the PR!

I think adding styles to the `Tooltip` component that depend on specific text should be avoided for the following reasons:

- The `Tooltip` component should not depend on a specific context
- The text may be localized and changed to a different text
- It also affects other tooltips that contain text such as `up` and `down`
- The tooltip is not guaranteed to appear at the bottom

The current Tooltip component [has a hardcoded gutter](https://github.com/WordPress/gutenberg/blob/7fcd7ea6ee5101d4548f6406bda25a51d36ad447/packages/components/src/tooltip/index.tsx#L133). My thought is that there may need to be a prop like `offset` or `tooltipOffest` to modify this gutter in some way, like in the `Popover` component.

cc @WordPress/gutenberg-components

Rishit30G commented:
Hey @t-hamano 
Thanks for reviewing the PR and sharing your valuable feedback, agree with all your points 🙇 
My intension was not to raise this PR but keep it as draft so that today I could have researched more on the caveats of my approach ( my bad ) 😅 
Nonetheless, I'll try to work on the approach you have shared 👍 
Thanks

tyxla commented on packages/components/src/tooltip/index.tsx:
```diff
@@ -128,7 +128,12 @@ function UnforwardedTooltip(
-					className={ clsx( 'components-tooltip', className ) }
+					className={ clsx(
+						'components-tooltip',
+						{ 'move-up': text?.toLowerCase().includes( 'up' ) },
+						{ 'move-down': text?.toLowerCase().includes( 'down' ) },
```
Inferring tooltip behavior or custom visual differences based on specific tooltip text doesn't make much sense to me. 

The problem is that `Tooltip` is a generic tooltip component, while `Move up" and "Move down" are coming from specific tooltip usages inside the block editor. 

With that in mind, if we're to add custom styles based on that, it shouldn't be in the generic component but should rather be in the consumer component.

tyxla replied:
Separately, I don't think it makes sense to use the text to infer particular behavior. What if we change or remove those text labels? Sounds fragile. Instead, we should try to do that from any other logical props, like `BlockMoverButton`'s `direction` for example.

tyxla commented on packages/components/src/tooltip/style.scss:
```diff
@@ -8,9 +8,14 @@
 	font-size: 12px;
 	padding: $grid-unit-05 $grid-unit-10;
 	z-index: z-index(".components-tooltip");
+	&.move-up {
+		margin-top: 25px;
+	}
+	&.move-down {
+		margin-top: 5px;
+	}
```
Those values seem arbitrary and if we were to add them, we should consider aligning them with the existing component system spacing.

tyxla reviewed (changes requested):
Unfortunately, I don't think this is the way to go.

I'd recommend trying to adjust the component that consumes the `Tooltip`, because this is a particular scenario related to a very specific use case.

Furthermore, if we were to add spacing to `Tooltip`s, I'd expect we do it in a consistent way that applies to all of them.

Rishit30G committed:
Function to get direction using direction and orientation

Rishit30G committed:
Add Tooltip direction

Rishit30G committed:
remove redundant styles

Rishit30G committed:
Alter gutter based on direction

Rishit30G committed:
fix tooltipDirection types

Rishit30G commented:
#### Updates on PR as per feedback 📝 

-  Updated the Button component with a new prop `tooltipDirection` 
- `tooltipDirection` can get a value of `up | down | right | left` based on the direction
-   Made a function `getDirection` which takes in two parameters `direction` and `orientation` which will help us avoid the problem of changing styles of right and left chevron labels when we only want to update `up` and `down` chevrons.
-   Updated the `Tooltip` with new prop `direction` (string) which will capture the value of `tooltipDirection` 
-   Used the `direction` prop in Tooltip component to set the `gutter` value for appropriate spacing 

cc: @tyxla @t-hamano 
 
Sharing the screencast for the same: 


https://github.com/user-attachments/assets/7760eef3-f2c6-41ef-9369-a9c5f0d0189e

mirka commented on packages/components/src/button/index.tsx:
```diff
@@ -110,6 +110,7 @@ export function UnforwardedButton(
 		children,
 		size = 'default',
 		text,
+		tooltipDirection,
```
Note that there is already a `tooltipPosition` prop on `Button` that can you can use to change the tooltip placement (e.g. `"top"` and `"bottom"`).

I don't think we will need any changes in the `@wordpress/components` package to land this PR. As for fine-tuning the "bottom" placement so the block mover down tooltip aligns with the other toolbar button tooltips, I think a cleaner strategy would be to fix the metrics of the block mover buttons so their vertical edges are aligned with the other buttons. If you add a visible outline to the buttons, you can see that the block mover buttons are shorter. This is causing the misalignment of the tooltips.

<img width="310" alt="Toolbar button metrics" src="https://github.com/user-attachments/assets/62e6bf52-fa71-4f4a-96d7-0d04b077fb60">

hbhalodia replied:
Hi @mirka, Agree on this, We just need to adjust this style added here which make this height - https://github.com/WordPress/gutenberg/blob/46f9fe0f31bf9ce5f6210cad96074206e90527e6/packages/block-editor/src/components/block-mover/style.scss#L12 as `24px`, by removing the `$grid-unit-05`, not sure why this added, may be to decrease the space between chevrons,

Also, need to adjust the top and bottom padding for `up` and `down` button to `10px`, so that it looks what it looked like earlier.

<img width="1440" alt="Screenshot 2024-08-14 at 10 14 22 AM" src="https://github.com/user-attachments/assets/ad0a3f8b-957f-441d-ab73-6e197ed1230d">

@Rishit30G, Could you please adjust this style and check if it helps, 

The `move up`, would still not in place, but `move down` would be consistent with different toolbar tooltips.

tyxla replied:
Agreed @mirka's suggestions are the way to go. We should use what we have and keep any specific customization outside of the `Tooltip` component.

Rishit30G replied:
Thanks for the suggestions @mirka, @hbhalodia, @tyxla 🙇 
- I have removed extra `tooltipDirection` and used `tooltipPosition` 
- Updated the style from `height: $block-toolbar-height * 0.5 - $grid-unit-05;` to `height: $block-toolbar-height * 0.5;`

t-hamano replied:
> the "$grid-unit-05", not sure why this added

This style was added when introducing a custom scrollbar to the block toolbar in #57444 (Sorry for not leaving a comment 🙇‍♂️).

If we remove `$grid-unit-05`, an unnatural shift will occur when the mover button is focused on Windows OS:

https://github.com/user-attachments/assets/c6e862f4-a3d7-45dd-8a5f-55966b6a920f

We tried a lot of different things to fix this problem in #57444, but for now, I think it's better not to remove `$grid-unit-05`.

Of course, if the top toolbar is disabled, there's no problem with removing `$grid-unit-05`, but if adjusting the tooltip position becomes complicated, I think it's better to leave `$grid-unit-05`.

Rishit30G replied:
Thanks for the info @t-hamano! 🙌🏻
I already made a workaround for this in previous commits, will see if I can incorporate it without breaking

Rishit30G replied:
Fixed ✅

ciampo replied:
Hey @Rishit30G , I agree with @mirka 's suggestions above, especially about

> I don't think we will need any changes in the `@wordpress/components` package to land this PR

Can we make sure this PR doesn't include changes to `Button` and `Tooltip` ? Thank you 🙏

Rishit30G committed:
fix block-mover height style

Rishit30G committed:
Update tooltipDirection with existing tooltipPosition

Rishit30G committed:
Update gutter value based on direction

Rishit30G committed:
revert grid-unit-05 style

Rishit30G committed:
Refactor block-mover button and tooltip gutter value
